src/zProposal.ts

```diff
@@ -0,0 +1,250 @@
+
+  get type() {
+    return this._type;
+  }
+
+  get author() {
+    return this._author;
+  }
+
+  get title() {
+    return this._title;
+  }
+
+  get body() {
+    return this._body;
+  }
+
+  get ipfs() {
+    return this._ipfs;
+  }
+
+  get choices() {
+    return this._choices;
+  }
+
+  get created() {
+    return this._created;
+  }
+
+  get start() {
+    return this._start;
+  }
+
+  get end() {
+    return this._end;
+  }
+
+  get state() {
+    return this._state;
+  }
+
+  get network() {
+    return this._network;
+  }
+
+  get snapshot() {
+    return this._snapshot;
+  }
+
+  get scores() {
+    return this._scores;
+  }
+
+  get votes() {
+    return this._votes;
+  }
+
+  get metadata() {
+    return this._metadata;
+  }
+
+  async getTokenMetadata(): Promise<TokenMetaData> {
+    if (!this._ipfs) {
+      throw Error(t('empty-voting-token'));
+    }
+    if (this._metadata) {
+      return this._metadata;
+    }
+
+    const ipfsData = await this._snapshotClient.ipfsGet(this._ipfs);
+    if (!ipfsData.data || !ipfsData.data.message) {
+      throw Error(t('empty-voting-token'));
+    }
+
+    const metadataJson = JSON.parse(ipfsData.data.message.metadata);
+    const sender = metadataJson.sender;
+    const recipient = metadataJson.recipient;
+    const token = metadataJson.token;
+    const decimals = metadataJson.decimals ?? 18;
+    const symbol = metadataJson.symbol ?? 'zToken';
+    const amount = metadataJson.amount;
+
+    this._metadata = {
+      sender,
+      recipient,
+      token,
+      decimals,
+      symbol,
+      amount,
+    };
+    return this._metadata;
+  }
+
+  async listVotes(from = 0, count = 30000, voter = ''): Promise<Vote[]> {
+    const votes = await this._snapshotClient.listVotes(
+      this._id,
+      from,
+      count,
+      voter
+    );
+
+    return votes.map((vote: any) => ({
+      voter: vote.voter,
+      choice: vote.choice as Choice,
+      power: vote.vp,
+    }));
+  }
+
+  async getVotingPowerOfUser(account: string): Promise<number> {
+    if (!this._metadata) {
+      throw Error(t('empty-metadata'));
+    }
+    return this._snapshotClient.getERCBalanceOf(
+      this._zDAO.zNA,
+      this._network,
+      parseInt(this._snapshot),
+      this._metadata.token,
+      this._metadata.decimals,
+      this._metadata.symbol,
+      account
+    );
+  }
+
+  async vote(signer: ethers.Wallet, choice: Choice): Promise<string> {
+    return this._snapshotClient.voteProposal(
+      signer,
+      this._zDAO.zNA,
+      this._id,
+      choice
+    );
+  }
+
+  async execute(signer: ethers.Wallet): Promise<TransactionResponse> {
+    const isOwner = await this._gnosisSafeClient.isOwnerAddress(
+      signer,
+      this._zDAO.zNA,
+      signer.address
+    );
+    if (!isOwner) {
+      throw Error(t('not-gnosis-owner'));
+    }
+
+    if (!this._metadata) {
+      throw Error(t('empty-metadata'));
+    }
+
+    if (!this._metadata?.token || this._metadata.token.length < 1) {
```

Comment:
At some point the ABI of the transaction that needs to be executed is going to be stored in the metadata itself (if it isn't already).

The gnosis safe client needs to be more generic.

Don't write the `gnosisSafeClient` to be able to send ERC20 tokens or Ether respectively, instead write it to execute *any* transaction. Ideally it has one method on it to execute `.proposeTransaction(...)` which would take the params
- safe address
- signer
- target (contract or recipient of Ether)
- abi (optional, since it may be just a transfer of Ether)
- params (optional, only needed when it's a contract execution, ie, abi is defined)
- value (amount of ether to send, optional as well)

Keep in mind that you can call a contract and send ether to it at the same time.


Also the params mentioned above should probably be put into an params interface and sent via object and not a list of parameters.